### PR TITLE
Add dynamic pod iface naming feature gate

### PR DESCRIPTION
**Pull request by qinqon**
### What this PR does
Before this PR:
The dynamic pod interface name discovery mechanism implemented at [1] was running unconditionally.

After this PR:
To enable the pod interface name discovery the feature gate related to it has to be enabled.

### Why we need it and why it was done in this way
The dynamic pod interface naming need to be well tested, with the feature gate is disabled by default and only activated at testing scenarios.

### Release note
```release-note
Add dynamic pod interface name feature gate
```

### CODE REVIEW

File: pkg/network/vmicontroller/status.go
Comment (orelmisan) [inline pkg/network/vmicontroller/status.go:42]: This is the beginning of the whole process, disable it and you've disabled everything.
```suggestion
if featureGateEnabled {
  interfaceStatuses = append(interfaceStatuses, calculatePrimaryIfaceStatus(clusterConfig, vmi, 
networkStatuses)...)
}
```

Comment (qinqon) [reply pkg/network/vmicontroller/status.go:42]: If we do here the interface will not be populated, we need to skip the PrimaryPodInterfaceName field to be set .

Comment (qinqon) [reply pkg/network/vmicontroller/status.go:42]: Nah  you right

File: tests/network/vmi_infosource.go
Comment (xpivarc) [inline tests/network/vmi_infosource.go:89]: I want to propose that we don't enable the feature gates unless needed to not pollute env. So can we rather enable this feature gate for the test suite and mark the tests with label so we are able to focus/skip?

Comment (EdDev) [reply tests/network/vmi_infosource.go:89]: We should enable the feature only on a specific test, preferably one that is already serial.

File: pkg/virt-handler/node-labeller/BUILD.bazel
Comment (qinqon) [reply pkg/virt-handler/node-labeller/BUILD.bazel:37]: I have just call `make generate`, there is `./pkg/virt-handler/node-labeller/capabilities_test.go` is weird that was not registered at BUILD.bazel before.

Comment (qinqon) [reply pkg/virt-handler/node-labeller/BUILD.bazel:37]: Ahh wait, didn't clean my repo :-/

Comment (qinqon) [reply pkg/virt-handler/node-labeller/BUILD.bazel]: Fixed

File: tests/network/vmi_infosource.go
Comment (qinqon) [reply tests/network/vmi_infosource.go]: Do I have to mark the test suite as serial ?

File: tests/network/vmi_infosource.go
Comment (EdDev) [inline tests/network/vmi_infosource.go:69]: This logic is uneeded.
Just enabled it, see [here](https://github.com/kubevirt/kubevirt/blob/d1a055b67783a852dbe1be519a6d32d0bee4c253/tests/network/bindingplugin.go#L50).

There is a test cleanup that restores KV to its original.

You also need to mark the test "serial" (see the ref one) because it touches the cluster level KV config.

File: pkg/network/setup/netstat.go
Comment (EdDev) [inline pkg/network/setup/netstat.go:117]: This should not be called when the FG is disabled.

File: pkg/network/setup/netconf.go
Comment (EdDev) [inline pkg/network/setup/netconf.go:116]: If the FG is not enabled, this should not be called.

File: pkg/virt-handler/non-root.go
Comment (EdDev) [inline pkg/virt-handler/non-root.go:138]: It would have been cleaner to pass in the network namescheme as input so the condition (about the FG) will occur at this level.

File: pkg/network/vmicontroller/status.go
Comment (qinqon) [reply pkg/network/vmicontroller/status.go:42]: I am adapting the unit test using a stub for clusterConfig but that's not compatible with the function firm expected by the vmi-controller, so I have create one function that uses the concrete type `virtconfig.ClusterConfig` that calls the function with the real logic and the one we are testing it.

The alternative is to use the concrete type `virtconfig.ClusterConfig` somehow at the test.

File: pkg/network/vmicontroller/status_test.go
Comment (qinqon) [reply pkg/network/vmicontroller/status_test.go:101]: no, do we need them ?

File: tests/decorators/decorators.go
Comment (qinqon) [reply tests/decorators/decorators.go:56]: I am labeling the test as suggested, so I have add it here, but I can harcoded directly at the test if feels better.

File: pkg/network/vmicontroller/status_test.go
Comment (EdDev) [reply pkg/network/vmicontroller/status_test.go:101]: We added a condition in the codebase, I guess we should at least check that path once.

File: tests/decorators/decorators.go
Comment (EdDev) [reply tests/decorators/decorators.go:56]: I do not see the need to label anything.
Did I miss a new standard or something?

File: pkg/virt-handler/non-root.go
Comment (qinqon) [reply pkg/virt-handler/non-root.go:138]: Don't know, this works.

File: pkg/network/vmicontroller/status.go
Comment (qinqon) [reply pkg/network/vmicontroller/status.go:42]: That do not work, golang fails stating that both interfaces are different, looks is only able to do duck typing with concrete types but not when comparing interfaces, meaning that you need a common interface in this case.

**Comment by qinqon** [comment]
/retest

not related
```
5:54:56: + '[' 1 -gt 1 ']'
05:54:56: ++ cat /kernel.args
05:54:56: ++ cat /additional.kernel.args
05:54:56: $++ cat /proc/sys/kernel/random/uuid
05:54:56: =+ exec qemu-system-x86_64 -enable-kvm -drive format=qcow2,file=/var/run/disk/disk.qcow2,if=virtio,cache=unsafe -drive format=qcow2,file=/var/run/disk/blockdev.qcow2,if=virtio,cache=unsafe -device virtio-net-pci,netdev=network0,mac=52:55:00:d1:55:01 -netdev tap,id=network0,ifname=tap01,script=no,downscript=no -device virtio-rng-pci -initrd /initrd.img -kernel /vmlinuz -append 'root=/dev/vda1 ro no_timer_check console=tty0 console=ttyS0,115200n8 net.ifnames=0 biosdevname=0 spectre_v2=off nopti intel_iommu=on modprobe.blacklist=nouveau  hugepagesz=2M hugepages=64' -vnc :01 -cpu host,migratable=no,+invtsc -m 11264M -smp 6 -serial pty -M q35,accel=kvm,kernel_irqchip=split -device intel-iommu,intremap=on,caching-mode=on -device intel-hda -device hda-duplex -device AC97 -uuid 556686e5-9211-4d0a-9f8c-cf133f64464a -serial pty
05:54:56: 5char device redirected to /dev/pts/1 (label serial0)
05:54:56: 5char device redirected to /dev/pts/2 (label serial1)
05:54:56: error executing kubeadm init --config /etc/kubernetes/kubeadm.conf -v5: failed to execute command: kubeadm init --config /etc/kubernetes/kubeadm.conf -v5
05:54:56: ./cluster-up/up.sh: line 34: pop_var_context: head of shell_variables not a function context
```

File: pkg/network/vmicontroller/status.go
Comment (EdDev) [reply pkg/network/vmicontroller/status.go:42]: You can use a simple lambda then.


Replace the [registration](https://github.com/kubevirt/kubevirt/blob/830e9fdfa86fbe3dc9ef604827bcec54b688000c/pkg/virt-controller/watch/application.go#L658) with this:
```go
func(clusterConfig *virtconfig.ClusterConfig, vmi *v1.VirtualMachineInstance, pod *k8sv1.Pod) error {
	return netvmicontroller.UpdateStatus(clusterConfig, vmi, pod)
},
```

File: pkg/network/vmicontroller/status.go
Comment (EdDev) [inline pkg/network/vmicontroller/status.go:73]: I do not really understand why we need this extra functions.

How about this (adding 3 lines):
```go
primaryPodIfaceName := multus.LookupPodPrimaryIfaceName(networkStatuses)
if primaryPodIfaceName == "" {
    primaryPodIfaceName = namescheme.PrimaryPodInterfaceName
}

if !clusterConfig.DynamicPodInterfaceNamingEnabled() {
    primaryPodIfaceName = ""
}
```

Originally, we have not treated the primary interface, and I think we should keep that protected by the FG.
That means we should have probably done something like this at a level up:
```go
if !clusterConfig.DynamicPodInterfaceNamingEnabled() {
    interfaceStatuses = append(interfaceStatuses, calculatePrimaryIfaceStatus(vmi, networkStatuses, clusterConfig)...)
}
```

File: pkg/network/setup/netconf.go
Comment (EdDev) [inline pkg/network/setup/netconf.go:127]: I prefer we do not inject this info down to the `netpod`, it should not care about this.

Can this be done?
```go
vmiIfacesStatuses := vmi.Status.Interfaces
if !cluterConfig.DynamicPodInterfaceNamingEnabled() {
    vmiIfacesStatuses = nil
}

netpod := netpod.NewNetPod(
    ...
    netpod.WithVMIIfaceStatuses(vmiIfacesStatuses),
)
```

File: pkg/network/setup/netstat.go
Comment (EdDev) [inline pkg/network/setup/netstat.go:126]: This should not be under the protection of the FG.

**Review by EdDev — CHANGES_REQUESTED** [review]

File: tests/network/vmi_infosource.go
Comment (qinqon) [reply tests/network/vmi_infosource.go:62]: Enabling FG at the proper context.

File: tests/network/vmi_infosource.go
Comment (qinqon) [reply tests/network/vmi_infosource.go:103]: Removing label.

File: pkg/network/vmicontroller/status.go
Comment (EdDev) [reply pkg/network/vmicontroller/status.go:73]: Sorry, I meant `if clusterConfig.DynamicPodInterfaceNamingEnabled()` not with a `!`.

File: pkg/network/setup/netconf.go
Comment (qinqon) [reply pkg/network/setup/netconf.go:127]: That was my idea from the beginning, but though It was not going to be well received.

**Comment by qinqon** [comment]
Manual test show VM without status interfaces 
```
apiVersion: v1
items:
- apiVersion: kubevirt.io/v1
  kind: VirtualMachineInstance
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"kubevirt.io/v1","kind":"VirtualMachineInstance","metadata":{"annotations":{},"name":"fedora","namespace":"default"},"spec":{"domain":{"devices":{"disks":[{"disk":{"bus":"virtio"},"name":"containerdisk"},{"disk":{"bus":"virtio"},"name":"cloudinitdisk"}],"rng":{}},"resources":{"requests":{"memory":"1024M"}}},"terminationGracePeriodSeconds":0,"volumes":[{"containerDisk":{"image":"quay.io/containerdisks/fedora:40"},"name":"containerdisk"},{"cloudInitNoCloud":{"userData":"#cloud-config\npassword: fedora\nchpasswd: { expire: False }"},"name":"cloudinitdisk"}]}}
      kubevirt.io/latest-observed-api-version: v1
      kubevirt.io/storage-observed-api-version: v1
    creationTimestamp: "2024-10-17T10:55:47Z"
    finalizers:
    - foregroundDeleteVirtualMachine
    generation: 13532
    labels:
      kubevirt.io/nodeName: ovn-worker2
    name: fedora
    namespace: default
    resourceVersion: "20545"
    uid: dc4fc560-7983-44b8-a43c-1860447f9719
  spec:
    architecture: amd64
    domain:
      cpu:
        cores: 1
        model: host-model
        sockets: 1
        threads: 1
      devices:
        disks:
        - disk:
            bus: virtio
          name: containerdisk
        - disk:
            bus: virtio
          name: cloudinitdisk
        interfaces:
        - bridge: {}
          name: default
        rng: {}
      features:
        acpi:
          enabled: true
      firmware:
        uuid: 3ceec7e4-682b-4c48-8fe6-11d177b684a6
      machine:
        type: q35
      memory:
        guest: 1024M
      resources:
        requests:
          memory: 1024M
    evictionStrategy: None
    networks:
    - name: default
      pod: {}
    terminationGracePeriodSeconds: 0
    volumes:
    - containerDisk:
        image: quay.io/containerdisks/fedora:40
        imagePullPolicy: IfNotPresent
      name: containerdisk
    - cloudInitNoCloud:
        userData: |-
          #cloud-config
          password: fedora
          chpasswd: { expire: False }
      name: cloudinitdisk
  status:
    activePods:
      4ba37752-b204-470e-849b-3f9bd592f16e: ovn-worker2
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2024-10-17T10:56:19Z"
      status: "True"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: null
      message: cannot migrate VMI which does not use masquerade, bridge with kubevirt.io/allow-pod-bridge-network-live-migration
        VM annotation or a migratable plugin to connect to the pod network
      reason: InterfaceNotLiveMigratable
      status: "False"
      type: LiveMigratable
    - lastProbeTime: null
      lastTransitionTime: null
      message: 'InterfaceNotLiveMigratable: cannot migrate VMI which does not use
        masquerade, bridge with kubevirt.io/allow-pod-bridge-network-live-migration
        VM annotation or a migratable plugin to connect to the pod network'
      reason: NotMigratable
      status: "False"
      type: StorageLiveMigratable
    - lastProbeTime: "2024-10-17T10:56:29Z"
      lastTransitionTime: null
      status: "True"
      type: AgentConnected
    currentCPUTopology:
      cores: 1
      sockets: 1
      threads: 1
    guestOSInfo:
      id: fedora
      kernelRelease: 6.8.5-301.fc40.x86_64
      kernelVersion: '#1 SMP PREEMPT_DYNAMIC Thu Apr 11 20:00:10 UTC 2024'
      machine: x86_64
      name: Fedora Linux
      prettyName: Fedora Linux 40 (Cloud Edition)
      version: 40 (Cloud Edition)
      versionId: "40"
    launcherContainerImageVersion: localhost:5000/virt-launcher@sha256:c592382723f2dd06ec7e258fa222d81cc997c6254bd1756a60f8e277b61e6f6a
    machine:
      type: pc-q35-rhel9.4.0
    memory:
      guestAtBoot: 1024M
      guestCurrent: 1000000Ki
      guestRequested: 1024M
    migrationMethod: BlockMigration
    migrationTransport: Unix
    nodeName: ovn-worker2
    phase: Running
    phaseTransitionTimestamps:
    - phase: Pending
      phaseTransitionTimestamp: "2024-10-17T10:55:47Z"
    - phase: Scheduling
      phaseTransitionTimestamp: "2024-10-17T10:55:47Z"
    - phase: Scheduled
      phaseTransitionTimestamp: "2024-10-17T10:56:19Z"
    - phase: Running
      phaseTransitionTimestamp: "2024-10-17T10:56:20Z"
    qosClass: Burstable
    runtimeUser: 107
    selinuxContext: none
    volumeStatus:
    - name: cloudinitdisk
      size: 1048576
      target: vdb
    - containerDiskVolume:
        checksum: 1480274367
      name: containerdisk
      target: vda
kind: List
metadata:
  resourceVersion: ""
```

**Comment by qinqon** [comment]
/retest

Cluster-up issues.

**Comment by vladikr** [comment]
/approve
/hold
looks fine to me. 
I'll let @xpivarc have a chance to look.

File: pkg/network/vmicontroller/status_test.go
Comment (xpivarc) [inline pkg/network/vmicontroller/status_test.go:180]: nit the `+` looks weird

File: pkg/network/vmicontroller/status_test.go
Comment (xpivarc) [inline pkg/network/vmicontroller/status_test.go:101]: Don't we want to have 2 cases one for enable and one for not?

File: pkg/network/vmicontroller/status_test.go
Comment (xpivarc) [inline pkg/network/vmicontroller/status_test.go:307]: For this one I would also assume you want fg off case

File: pkg/network/vmicontroller/status_test.go
Comment (xpivarc) [inline pkg/network/vmicontroller/status_test.go:93]: Generally I would use a config with default as `off` and then use the other one only for test cases introduced in #13018

File: pkg/network/setup/netconf_test.go
Comment (xpivarc) [inline pkg/network/setup/netconf_test.go:61]: Where do we test `off` case?

File: pkg/network/setup/netstat_test.go
Comment (xpivarc) [inline pkg/network/setup/netstat_test.go:856]: I would default to `off` as that is the default and use `on`  when you need (one test case?)

File: tests/network/vmi_infosource.go
Comment (xpivarc) [inline tests/network/vmi_infosource.go:80]: How do we cover these cases without fg?

**Comment by EdDev** [comment]
I acknowledge tge test coverage with the new FG is not perfect, but this FG requirement came to us by surprise

I do not think we have the capacity to do this by the book before the FF.
Most of the e2e tests do not use the FG, so we are protected.

When I approved the network code I took this risk.
If new code is added, I prefer sig-network members will review it. But we will not have them until after the FF.

**Comment by vladikr** [comment]
> I acknowledge tge test coverage with the new FG is not perfect, but this FG requirement came to us by surprise
> 
> I do not think we have the capacity to do this by the book before the FF. Most of the e2e tests do not use the FG, so we are protected.
> 
> When I approved the network code I took this risk. If new code is added, I prefer sig-network members will review it. But we will not have them until after the FF.

Isn't the FF on the 22nd? 
Why is adding a test covering the `off` case so hard?

**Comment by xpivarc** [comment]
@EdDev 
I am not sure if you understand my comments. My comments were mainly about the unit tests which we run with the feature gate enabled. This is contradicting the state we have, the feature being off by default.  I don't think this is major overhaul and I can provide quick reviews...

File: pkg/network/vmicontroller/status_test.go
Comment (qinqon) [reply pkg/network/vmicontroller/status_test.go]: I have adapt the tests to use off and on where needed.

File: pkg/network/vmicontroller/status_test.go
Comment (xpivarc) [inline pkg/network/vmicontroller/status_test.go:190]: `dynamicPodInterfaceNamingDisabled`

File: pkg/network/setup/netconf_test.go
Comment (qinqon) [reply pkg/network/setup/netconf_test.go]: Done

File: pkg/network/setup/netstat_test.go
Comment (qinqon) [reply pkg/network/setup/netstat_test.go]: Done

File: pkg/network/vmicontroller/status_test.go
Comment (xpivarc) [inline pkg/network/vmicontroller/status_test.go:383]: For this case I think we need variations (on & off).

**Comment by vladikr** [comment]
Thanks @qinqon 
/unhold

**Pull request merged**
